tools/ts_library_builder/build_library.ts

```diff
@@ -0,0 +1,453 @@
+        .getFilePath()
+        .replace(/\.ts$/, ".d.ts");
+      const dtsSourceFile = declarationProject.getSourceFileOrThrow(
+        dtsFilePath
+      );
+      mergedModule.addStatements(
+        namespaceSourceFile(dtsSourceFile, {
+          debug,
+          namespace: declaration.getNamespaceImportOrThrow().getText(),
+          rootPath: basePath,
+          sourceFileMap
+        })
+      );
+    }
+  }
+
+  if (debug) {
+    addSourceComment(mergedModule, sourceFile, basePath);
+  }
+}
+
+/**
+ * Generate the runtime library for Deno and write it to the supplied out file
+ * name.
+ */
+export function main({
+  basePath,
+  buildPath,
+  debug,
+  outFile,
+  silent
+}: BuildLibraryOptions) {
+  if (!silent) {
+    console.log("-----");
+    console.log("build_lib");
+    console.log();
+    console.log(`basePath: "${basePath}"`);
+    console.log(`buildPath: "${buildPath}"`);
+    console.log(`debug: ${!!debug}`);
+    console.log(`outFile: "${outFile}"`);
+    console.log();
+  }
+
+  // the inputProject will take in the TypeScript files that are internal
+  // to Deno to be used to generate the library
+  const inputProject = new Project({
+    compilerOptions: {
+      baseUrl: basePath,
+      declaration: true,
+      emitDeclarationOnly: true,
+      lib: [],
+      module: ModuleKind.AMD,
+      moduleResolution: ModuleResolutionKind.NodeJs,
+      noLib: true,
+      paths: {
+        "*": ["*", `${buildPath}/*`]
+      },
+      preserveConstEnums: true,
+      strict: true,
+      stripInternal: true,
+      target: ScriptTarget.ESNext
+    }
+  });
+
+  // Add the input files we will need to generate the declarations, `globals`
+  // plus any modules that are importable in the runtime need to be added here
+  // plus the `lib.esnext` which is used as the base library
+  inputProject.addExistingSourceFiles([
+    `${basePath}/node_modules/typescript/lib/lib.esnext.d.ts`,
+    `${basePath}/js/deno.ts`,
+    `${basePath}/js/globals.ts`
+  ]);
+
+  // emit the project, which will be only the declaration files
+  const inputEmitResult = inputProject.emitToMemory();
+
+  // the declaration project will be the target for the emitted files from
+  // the input project, these will be used to transfer information over to
+  // the final library file
+  const declarationProject = new Project({
+    compilerOptions: {
+      baseUrl: basePath,
+      moduleResolution: ModuleResolutionKind.NodeJs,
+      noLib: true,
+      paths: {
+        "*": ["*", `${buildPath}/*`]
+      },
+      strict: true,
+      target: ScriptTarget.ESNext
+    },
+    useVirtualFileSystem: true
+  });
+
+  // we don't want to add to the declaration project any of the original
+  // `.ts` source files, so we need to filter those out
+  const jsPath = join(basePath, "js");
+  const inputProjectFiles = inputProject
+    .getSourceFiles()
+    .map(sourceFile => sourceFile.getFilePath())
+    .filter(filePath => !filePath.startsWith(jsPath));
```

Comment:
@kitsonk Was just thinking about this. I think this would work?

```ts
const jsDir = inputProject.getDirectoryOrThrow(join(basePath, "js"));
const inputProjectFiles = jsDir.getDescendantSourceFiles();
```

Reply:
Oh, whoops! It's the opposite of that. Try this:

```ts
const jsDir = inputProject.getDirectoryOrThrow(join(basePath, "js"));
const inputProjectFiles = inputProject
    .getSourceFiles()
    .filter(sourceFile => !jsDir.isAncestorOf(sourceFile));
```

Reply:
What I did in the end for now was just create a `normalizePath()` that replicates the TypeScript functionality (before I really took a look at this comment)...  It looks like it is finally passing CI.